### PR TITLE
Conv3d new

### PR DESCRIPTION
**Pull request by j4yan**
3D convolution: 
- [x] Added `DeviceConv3dFwdNaive_Input_N_Di_Hi_Wi_C_Weight_K_Z_Y_X_C_Output_N_Do_Ho_Wo_K`. This implementation doesn't use GridwiseBatchedGemm. Instead, the passed-in pointers, p_a_grid and p_c_grid, are shifted based on the batches that a wave-group falls in.
- [x] Added a naive conv3d implementation, `DeviceConv3dFwdNaive_Input_N_Di_Hi_Wi_C_Weight_K_Z_Y_X_C_Output_N_Do_Ho_Wo_K` for verification.
- [x] Added a template paraemter, `Block2CTileMap` to `GridwiseGemm::Run`. The old `GridwiseGemm::Block2CTileMap` is renamed to `DefaultBlock2CTileMap`.
- [x] Moved overloaded arithmetic operators of integral_constant/Number to integral_constant, since the implementations covers more types.
- [x] Added a new transform primitive, `Modulo`. It's not efficient because of the modulo operations involved.  Also it's not used in current implementation, so can be safely removed.
- [x] Fixed some coding style issues; renamed some variables for consistency (see variables with name `c_blockid_to*`).

### CODE REVIEW

File: composable_kernel/include/tensor_description/multi_index_transform.hpp [hunk 1]
```diff
@@ -1862,5 +1862,92 @@ struct Slice
...
+    }
+
+    __host__ __device__ static constexpr index_t GetNumOfLowerDimension() { return 1; }
+
+    __host__ __device__ static constexpr index_t GetNumOfUpperDimension() { return 1; }
+
+    __host__ __device__ constexpr const auto& GetUpperLengths() const { return up_lengths_; }
+
+    template <typename LowIdx, typename UpIdx>
+    __host__ __device__ constexpr void CalculateLowerIndex(LowIdx& idx_low,
```
Comment (asroy) [inline composable_kernel/include/tensor_description/multi_index_transform.hpp:1893]: This is naive implementation, need to change in future PR

File: composable_kernel/include/tensor_description/multi_index_transform.hpp [hunk 2]
```diff
@@ -1862,5 +1862,92 @@ struct Slice
...
+
+        idx_low(Number<0>{}) = idx_up[Number<0>{}] % modulus_;
+    }
+
+    template <typename LowIdxDiff,
+              typename UpIdxDiff,
+              typename LowIdx,
+              typename UpIdx,
+              index_t Hack>
+    __host__ __device__ void UpdateLowerIndex(LowIdxDiff& idx_diff_low,
```
Comment (asroy) [inline composable_kernel/include/tensor_description/multi_index_transform.hpp:1907]: This is naive implementation, need to change in future PR

File: composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp [hunk 3]
```diff
@@ -53,6 +53,63 @@ __global__ void
...
+                                                  p_shared,
+                                                  a_grid_desc_k0_m_k1,
+                                                  b_grid_desc_k0_n_k1,
+                                                  c_grid_desc_m0_n0_m1_n1_m2_m3_m4_n2,
+                                                  a_element_op,
+                                                  b_element_op,
+                                                  c_element_op,
+                                                  block_2_ctile_map);
+}
 #elif CK_EXPERIMENTAL_PASS_TENSOR_DESCRIPTOR_BY_VOID_POINTER
```
Comment (asroy) [inline composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp:113]: This logic "CK_EXPERIMENTAL_PASS_TENSOR_DESCRIPTOR_BY_VOID_POINTER" is no longer needed. You can remove it

File: composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp [hunk 4]
```diff
@@ -53,6 +53,63 @@ __global__ void
...
+          typename BGridDesc_K0_N_K1,
+          typename CGridDesc_M0_N0_M1_N1_M2_M3_M4_N2,
+          typename AElementwiseOperation,
+          typename BElementwiseOperation,
+          typename CElementwiseOperation,
+          typename Block2CTileMap,
+          bool HasMainKBlockLoop>
+__global__ void
+#if CK_USE_LAUNCH_BOUNDS
+    __launch_bounds__(CK_MAX_THREAD_PER_BLOCK, CK_MIN_BLOCK_PER_CU)
```
Comment (asroy) [inline composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp:70]: I think you can move this function into DeviceOp for your conv3d

File: composable_kernel/include/utility/array.hpp [hunk 5]
```diff
@@ -49,7 +49,7 @@ template <typename X, typename... Xs>
 __host__ __device__ constexpr auto make_array(X&& x, Xs&&... xs)
 {
     using data_type = remove_cvref_t<X>;
-    return Array<data_type, sizeof...(Xs) + 1>{{std::forward<X>(x), std::forward<Xs>(xs)...}};
+    return Array<data_type, sizeof...(Xs) + 1>{std::forward<X>(x), std::forward<Xs>(xs)...};
```
Comment (asroy) [inline composable_kernel/include/utility/array.hpp:52]: This fixes an error?

File: composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp (hunk 4 above)
Comment (j4yan) [reply composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp:70]: Makes sense. Will do.

File: composable_kernel/include/utility/array.hpp (hunk 5 above)
Comment (j4yan) [reply composable_kernel/include/utility/array.hpp:52]: It doesn't. But why using double brackets?

File: composable_kernel/include/utility/integral_constant.hpp [hunk 6]
```diff
@@ -13,5 +13,43 @@ struct integral_constant
     __host__ __device__ constexpr value_type operator()() const noexcept { return value; }
 };
 
+template <typename Index0, Index0 X, typename Index1, Index1 Y>
```
Comment (asroy) [inline composable_kernel/include/utility/integral_constant.hpp:16]: template <typename TX, Index0 VX, typename TY, Index1 VY>

File: device_operation/include/device_conv3d_fwd_xdl_ndhwc_kzyxc_ndhwk.hpp [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+    using CDataType = OutDataType;
+    // TODO make A/B datatype different
+    using ABDataType = InDataType;
+
+    static constexpr auto I0 = Number<0>{};
+    static constexpr auto I1 = Number<1>{};
+    static constexpr auto I2 = Number<2>{};
+    static constexpr auto I3 = Number<3>{};
+
+    static index_t GetMaxAllowableSubBatchSize(const index_t N,
```
Comment (asroy) [inline device_operation/include/device_conv3d_fwd_xdl_ndhwc_kzyxc_ndhwk.hpp:72]: Add comment in the code to explain why you need to split

**Review by asroy — CHANGES_REQUESTED** [review]
Need some cosmetic changes

File: composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp (hunk 3 above)
Comment (j4yan) [reply composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp:113]: what about CK_EXPERIMENTAL_STATIC_TENSOR_DESCRIPTOR?

**Review by asroy — DISMISSED** [review]

File: device_operation/include/device_conv3d_fwd_xdl_ndhwc_kzyxc_ndhwk.hpp (hunk 7 above)
Comment (j4yan) [reply device_operation/include/device_conv3d_fwd_xdl_ndhwc_kzyxc_ndhwk.hpp:134]: comments added.

File: composable_kernel/include/utility/integral_constant.hpp (hunk 6 above)
Comment (j4yan) [reply composable_kernel/include/utility/integral_constant.hpp]: done.

**Review by zjing14 — DISMISSED** [review]
LGTM

**Review by asroy — APPROVED** [review]

**Pull request merged**